### PR TITLE
[FLINK-33463][Connector/JDBC] Support the implementation of dynamic source tables based on the new source

### PR DESCRIPTION
**Pull request by RocMarshal**
- Support the implementation of dynamic source tables based on the new source

### CODE REVIEW

**Comment by RocMarshal** [comment]
blocked by https://github.com/apache/flink-connector-jdbc/pull/116

**Comment by RocMarshal** [comment]
HI, @eskabetxe @caicancai  Could you help have a review if you had the free time ? thx a lot.

File: flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java [hunk 1]
```diff
@@ -144,7 +146,8 @@ void testJdbcReadProperties() {
                         readOptions,
                         LookupOptions.MAX_RETRIES.defaultValue(),
                         null,
-                        SCHEMA.toPhysicalRowDataType());
+                        SCHEMA.toPhysicalRowDataType(),
+                        FactoryMocks.IDENTIFIER.asSummaryString());
```
Comment (eskabetxe) [inline flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java:150]: could we test in some way the source naming?

File: flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java [hunk 2]
```diff
@@ -179,13 +205,34 @@ public ScanRuntimeProvider getScanRuntimeProvider(ScanContext runtimeProviderCon
...
-        builder.setRowDataTypeInfo(
-                runtimeProviderContext.createTypeInformation(physicalRowDataType));
+        builder.setResultExtractor(new RowDataResultExtractor(dialect.getRowConverter(rowType)));
+        builder.setTypeInformation(scanContext.createTypeInformation(physicalRowDataType));
+        options.getProperties()
+                .forEach(
+                        (key, value) ->
+                                builder.setConnectionProperty(key.toString(), value.toString()));
+        JdbcSource<RowData> source = builder.build();
+        return new DataStreamScanProvider() {
```
[the remark this answers is not in the document]
Comment (RocMarshal) [reply flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java]: updated.

File: flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java (hunk 1 above)
Comment (RocMarshal) [reply flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java:158]: image
image


I added the corresponding cases for testing.
As shown in the screenshot pics, The name is overwrite by the translation mechanism.
Maybe we should drop the `tableIdentifier` attribute at present ?

**Review by RocMarshal — COMMENTED** [review]
Thanks @eskabetxe for the review.
I made some change based on your comments.
PTAL if you had the free time~ :)

File: flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java (hunk 1 above)
Comment (eskabetxe) [reply flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java:158]: If is overwrite I would say yes, we should drop the attribute, as at the end is not used..

Comment (RocMarshal) [reply flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableFactoryTest.java]: @eskabetxe updated!   :)

**Review by eskabetxe — APPROVED** [review]
LGTM

**Review by RocMarshal — COMMENTED** [review]
Thanks @eskabetxe for the quick review~
now ping @1996fanrui

**Review by 1996fanrui — APPROVED** [review]
Thanks @RocMarshal for the contribution, and @eskabetxe for the hard review!

LGTM

**Comment by RocMarshal** [comment]
Hi, @1996fanrui .
Kafka introduced KafkaSource API in the Flink 1.12 branch https://issues.apache.org/jira/browse/FLINK-18323  
Afterwards, in Flink version 1.14, KafkaDynamicTableSource was implemented based on KafkaSource  https://issues.apache.org/jira/browse/FLINK-22914
 
So, Maybe we should revert this commit.
Please let me know what's your opinion.
CC @eskabetxe

**Comment by 1996fanrui** [comment]
> Maybe we should revert this commit and re-commit it in the next second or third version

Sounds make sense to me. The JdbcSource is introduced in the current version, it may be not stable. So it's better to consider it as the default Jdbc connector for SQL job after community think it's stable.

**Pull request merged**